File: internal/output/formatter_test.go
package output

import (
	"bytes"
	"testing"
	"time"

	"cli-debugger/pkg/types"
)

func TestNewTextFormatter(t *testing.T) {
	f := NewTextFormatter(true)
	if f == nil {
		t.Error("TextFormatter 不应为 nil")
	}
	if !f.color {
		t.Error("color 应该为 true")
	}
}

func TestNewJSONFormatter(t *testing.T) {
	f := NewJSONFormatter()
	if f == nil {
		t.Error("JSONFormatter 不应为 nil")
	}
}

func TestNewTableFormatter(t *testing.T) {
	f := NewTableFormatter(true)
	if f == nil {
		t.Error("TableFormatter 不应为 nil")
	}
}

func TestGetFormatterType(t *testing.T) {
	tests := []struct {
		input    string
		expected FormatterType
	}{
		{"text", TextFormatter},
		{"json", JSONFormatter},
		{"table", TableFormatter},
		{"invalid", TextFormatter}, // 默认值
		{"", TextFormatter},        // 空字符串默认值
	}

	for _, tt := range tests {
		result := GetFormatterType(tt.input)
		if result != tt.expected {
			t.Errorf("GetFormatterType(%q) = %v, expected %v", tt.input, result, tt.expected)
		}
	}
}

func TestNewFormatter(t *testing.T) {
	tests := []struct {
		typeStr  string
		expected FormatterType
	}{
		{"text", TextFormatter},
		{"json", JSONFormatter},
		{"table", TableFormatter},
	}

	for _, tt := range tests {
		f := NewFormatter(tt.typeStr, true)
		if f == nil {
			t.Errorf("NewFormatter(%q) 返回 nil", tt.typeStr)
		}
	}
}

func TestTextFormatter_FormatVersion(t *testing.T) {
	var buf bytes.Buffer
	f := NewTextFormatter(false)
	f.SetWriter(&buf)

	info := &types.VersionInfo{
		ProtocolVersion: "1.8",
		RuntimeVersion:  "17.0.8",
		RuntimeName:     "OpenJDK",
		Description:     "Test JVM",
	}

	err := f.FormatVersion(info)
	if err != nil {
		t.Errorf("FormatVersion 失败：%v", err)
	}

	output := buf.String()
	if len(output) == 0 {
		t.Error("输出不应为空")
	}
}

func TestTextFormatter_FormatThreads(t *testing.T) {
	var buf bytes.Buffer
	f := NewTextFormatter(false)
	f.SetWriter(&buf)

	threads := []*types.ThreadInfo{
		{ID: "1", Name: "main", State: "running", Priority: 5},
		{ID: "2", Name: "worker", State: "waiting", Priority: 3, IsDaemon: true},
	}

	err := f.FormatThreads(threads)
	if err != nil {
		t.Errorf("FormatThreads 失败：%v", err)
	}

	output := buf.String()
	if len(output) == 0 {
		t.Error("输出不应为空")
	}
}

func TestTextFormatter_FormatStack(t *testing.T) {
	var buf bytes.Buffer
	f := NewTextFormatter(false)
	f.SetWriter(&buf)

	frames := []*types.StackFrame{
		{ID: "1", Method: "main", Location: "com.example.Main", Line: 10},
		{ID: "2", Method: "run", Location: "com.example.Worker", Line: 25, IsNative: true},
	}

	err := f.FormatStack(frames)
	if err != nil {
		t.Errorf("FormatStack 失败：%v", err)
	}

	output := buf.String()
	if len(output) == 0 {
		t.Error("输出不应为空")
	}
}

func TestTextFormatter_FormatVariables(t *testing.T) {
	var buf bytes.Buffer
	f := NewTextFormatter(false)
	f.SetWriter(&buf)

	variables := []*types.Variable{
		{Name: "count", Type: "int", Value: 42, IsPrimitive: true},
		{Name: "name", Type: "String", Value: "test", IsPrimitive: false},
		{Name: "active", Type: "boolean", Value: true, IsPrimitive: true},
	}

	err := f.FormatVariables(variables)
	if err != nil {
		t.Errorf("FormatVariables 失败：%v", err)
	}

	output := buf.String()
	if len(output) == 0 {
		t.Error("输出不应为空")
	}
}

func TestTextFormatter_FormatBreakpoints(t *testing.T) {
	var buf bytes.Buffer
	f := NewTextFormatter(false)
	f.SetWriter(&buf)

	breakpoints := []*types.BreakpointInfo{
		{ID: "1", Location: "com.example.Main:10", Enabled: true, HitCount: 5},
		{ID: "2", Location: "com.example.Worker:25", Enabled: false, Condition: "x > 10"},
	}

	err := f.FormatBreakpoints(breakpoints)
	if err != nil {
		t.Errorf("FormatBreakpoints 失败：%v", err)
	}

	output := buf.String()
	if len(output) == 0 {
		t.Error("输出不应为空")
	}
}

func TestTextFormatter_FormatEvent(t *testing.T) {
	var buf bytes.Buffer
	f := NewTextFormatter(false)
	f.SetWriter(&buf)

	event := &types.DebugEvent{
		Type:      "breakpoint",
		ThreadID:  "1",
		Location:  "com.example.Main:10",
		Timestamp: time.Now(),
		Data:      map[string]interface{}{"reason": "user breakpoint"},
	}

	err := f.FormatEvent(event)
	if err != nil {
		t.Errorf("FormatEvent 失败：%v", err)
	}

	output := buf.String()
	if len(output) == 0 {
		t.Error("输出不应为空")
	}
}

func TestJSONFormatter_FormatVersion(t *testing.T) {
	var buf bytes.Buffer
	f := NewJSONFormatter()
	f.SetWriter(&buf)

	info := &types.VersionInfo{
		ProtocolVersion: "1.8",
		RuntimeVersion:  "17.0.8",
		RuntimeName:     "OpenJDK",
	}

	err := f.FormatVersion(info)
	if err != nil {
		t.Errorf("FormatVersion 失败：%v", err)
	}

	output := buf.String()
	if len(output) == 0 {
		t.Error("输出不应为空")
	}
}

File: src/output/__tests__/formatter.test.ts
import { describe, it, expect } from "vitest";
import { Writable } from "node:stream";
import { TextFormatter } from "../text.js";
import { JsonFormatter } from "../json.js";
import { TableFormatter } from "../table.js";
import { createFormatter } from "../index.js";
import type { ThreadInfo, StackFrame, VersionInfo } from "../../types/index.js";

// Mock writable stream for testing
class MockWritable extends Writable {
  public output: string = "";

  _write(chunk: Buffer, encoding: string, callback: () => void): void {
    this.output += chunk.toString();
    callback();
  }

  clear(): void {
    this.output = "";
  }
}

describe("Output Formatters", () => {
  const mockWriter = new MockWritable();

  const mockThreads: ThreadInfo[] = [
    { id: 1, name: "main", state: "RUNNABLE", isSuspended: false, priority: 5 },
    { id: 2, name: "worker", state: "WAITING", isSuspended: true, priority: 5 },
  ];

  const mockFrames: StackFrame[] = [
    {
      method: "main",
      location: "Main.java",
      line: 10,
      isNative: false,
    },
    {
      method: "run",
      location: "Worker.java",
      line: 25,
      isNative: true,
    },
  ];

  const mockVersion: VersionInfo = {
    protocolVersion: "1.0",
    runtimeName: "Java",
    runtimeVersion: "17",
    description: "Test VM",
  };

  describe("TextFormatter", () => {
    const formatter = new TextFormatter({ color: false });
    formatter.setWriter(mockWriter);

    it("should format threads", async () => {
      mockWriter.clear();
      await formatter.formatThreads(mockThreads);
      expect(mockWriter.output).toContain("Thread 1: main");
      expect(mockWriter.output).toContain("Thread 2: worker");
      expect(mockWriter.output).toContain("[RUNNING]");
      expect(mockWriter.output).toContain("[SUSPENDED]");
    });

    it("should format stack frames", async () => {
      mockWriter.clear();
      await formatter.formatStack(mockFrames);
      expect(mockWriter.output).toContain("#0 main at Main.java:10");
      expect(mockWriter.output).toContain("#1 run at Worker.java:25");
      expect(mockWriter.output).toContain("[native]");
    });

    it("should format version info", async () => {
      mockWriter.clear();
      await formatter.formatVersion(mockVersion);
      expect(mockWriter.output).toContain("Protocol: 1.0");
      expect(mockWriter.output).toContain("Runtime: Java 17");
      expect(mockWriter.output).toContain("Description: Test VM");
    });

    it("should format errors", async () => {
      mockWriter.clear();
      const error = new Error("Test error");
      await formatter.formatError(error);
      expect(mockWriter.output).toContain("Error: Test error");
    });
  });

  describe("JsonFormatter", () => {
    const formatter = new JsonFormatter();
    formatter.setWriter(mockWriter);

    it("should format threads as JSON", async () => {
      mockWriter.clear();
      await formatter.formatThreads(mockThreads);
      const output = JSON.parse(mockWriter.output);
      expect(output.type).toBe("threads");
      expect(output.data).toHaveLength(2);
      expect(output.data[0].name).toBe("main");
    });

    it("should format stack frames as JSON", async () => {
      mockWriter.clear();
      await formatter.formatStack(mockFrames);
      const output = JSON.parse(mockWriter.output);
      expect(output.type).toBe("stack");
      expect(output.data).toHaveLength(2);
      expect(output.data[0].method).toBe("main");
    });

    it("should format version as JSON", async () => {
      mockWriter.clear();
      await formatter.formatVersion(mockVersion);
      const output = JSON.parse(mockWriter.output);
      expect(output.type).toBe("version");
      expect(output.data.protocolVersion).toBe("1.0");
    });

    it("should format errors as JSON", async () => {
      mockWriter.clear();
      const error = new Error("Test error");
      await formatter.formatError(error);
      const output = JSON.parse(mockWriter.output);
      expect(output.type).toBe("error");
      expect(output.data.message).toBe("Test error");
    });
  });

  describe("TableFormatter", () => {
    const formatter = new TableFormatter({ color: false });
    formatter.setWriter(mockWriter);

    it("should format threads as table", async () => {
      mockWriter.clear();
      await formatter.formatThreads(mockThreads);
      expect(mockWriter.output).toContain("ID");
      expect(mockWriter.output).toContain("Name");
      expect(mockWriter.output).toContain("State");
      expect(mockWriter.output).toContain("main");
      expect(mockWriter.output).toContain("worker");
    });

    it("should format stack frames as table", async () => {
      mockWriter.clear();
      await formatter.formatStack(mockFrames);
      expect(mockWriter.output).toContain("#");
      expect(mockWriter.output).toContain("Method");
      expect(mockWriter.output).toContain("Location");
      expect(mockWriter.output).toContain("main");
      expect(mockWriter.output).toContain("run");
    });

    it("should format version as table", async () => {
      mockWriter.clear();
      await formatter.formatVersion(mockVersion);
      expect(mockWriter.output).toContain("Property");
      expect(mockWriter.output).toContain("Value");
      expect(mockWriter.output).toContain("Protocol");
      expect(mockWriter.output).toContain("1.0");
    });
  });

  describe("createFormatter factory", () => {
    it("should create text formatter", () => {
      const formatter = createFormatter({ type: "text" });
      expect(formatter).toBeInstanceOf(TextFormatter);
    });

    it("should create json formatter", () => {
      const formatter = createFormatter({ type: "json" });
      expect(formatter).toBeInstanceOf(JsonFormatter);
    });

    it("should create table formatter", () => {
      const formatter = createFormatter({ type: "table" });
      expect(formatter).toBeInstanceOf(TableFormatter);
    });

    it("should set writer when provided", () => {
      const formatter = createFormatter({ type: "text", writer: mockWriter });
      expect(formatter).toBeInstanceOf(TextFormatter);
    });
  });
});
